sdcm/utils/k8s.py

```diff
@@ -609,6 +614,251 @@ def stop(self, timeout=None):
+            if add_pod_name_as_kwarg:
+                kwargs['pod_name'] = pod_name
+            # TODO: make following be 'debug' message
+            LOGGER.info("Calling '%s' callback %s", func.__name__, suffix)
+            try:
+                func(*args, **kwargs)
+            except Exception as exc:  # pylint: disable=broad-except
+                LOGGER.warning("Callback call failed %s: %s", suffix, str(exc))
+
+        data = self.mapper_dict.get(namespace, {})
+        for dict_key in (pod_name, '__each__'):
+            callbacks = data.get(dict_key, {}).get('callbacks', [])
+            if dict_key == '__each__':
+                # NOTE: when 'add_pod_name_as_kwarg' is True it means
+                #       we run a pod-specific function
+                #       and it also means we should run it before namespace-specific ones.
+                callbacks = sorted(callbacks, key=lambda c: c[-1], reverse=True)
+            for callback in callbacks:
+                process_callback(
+                    callback=callback[:-1], namespace=namespace, pod_name=pod_name,
+                    add_pod_name_as_kwarg=callback[-1])
+
+    def register_callbacks(self, callbacks: list[Callable],
+                           namespace: str, pod_name: str = '__each__',
+                           add_pod_name_as_kwarg: bool = False) -> None:
+        """Register callbacks to be called after a Scylla pod IP change.
+
+        Callbacks may be of the following types:
+        - called after a specific Scylla pod IP change
+        - called after any Scylla pod IP change in a specific namespace
+        - called after any Scylla pod IP change in a specific namespace
+          getting 'pod_name=%pod_name%' kwarg.
+          Useful for the objects which will be created later in the future.
+
+        Examples of usages:
+
+        # Per-pod
+        - register_callbacks(func1,
+                             namespace=namespace, pod_name='pod-name-1')
+        - register_callbacks([[func2, args, kwargs]],
+                             namespace=namespace, pod_name='pod-name-2')
+        - register_callbacks([func3, [func3, args2, kwargs2]],
+                             namespace=namespace, pod_name='pod-name-3')
+
+        # Per-pod but with automatic addition of a pod name as a kwarg where the key is 'pod_name'
+        - register_callbacks(func,
+                             namespace=namespace,
+                             add_pod_name_as_kwarg=True)
+
+        # Per-namespace
+        - register_callbacks([[func, args, kwargs], [func2, args, kwargs]],
+                             namespace=namespace)
+
+        Per-pod callbacks get run first, then per-namespace ones.
+        """
+        if not callbacks:
+            LOGGER.warning(
+                "No callbacks are provided. Nothing to register. "
+                "namespace='%s', pod_name='%s'",
+                namespace, pod_name)
+            return
+        elif callable(callbacks):
```

Comment:
if you support both a list, and one item, it should marked in the type annotations as such

Reply:
ok, will do

Reply:
Done